docs/en/050_Configuring_The_Arsse.md
## General settings

### lang

| Type   | Default |
|--------|---------|
| string | `"en"`  |

The default language locale, mostly used when reporting errors on the command line or in logs. Currently only `"en"` (English) is available.

## Database settings

### dbDriver

| Type   | Default     |
|--------|-------------|
| string | `"sqlite3"` |

The database driver to use. The following values are understood:

- `"sqlite3"` for SQLite databases
- `"postgresql"` for PostgreSQL databases
- `"mysql"` for MySQL databases

It is also possible to specify the fully-qualified name of a class which implements the database driver interface. For example, specifying `"JKingWeb\Arsse\Db\SQLite3\PDODriver"` would use the PDO driver for SQLite 3.

### dbAutoUpdate

| Type    | Default |
|---------|---------|
| boolean | `true`  |

Whether to attempt to automatically upgrade the database schema when upgrading The Arsse to a new version with schema changes.

If set to `false`, the database schema must be manually upgraded. Schema files can be found under `sql/<backend>/#.sql`; the `UPGRADING` file will advise when a schema upgrade is required.

### dbTimeoutConnect

| Type             | Default |
|------------------|---------|
| interval or null | `5.0`   |

The number of seconds to wait before returning a timeout error when connecting to a database server. The special value `null` waits the maximum amount of time, while `0.0` waits the minimum amount of time. The minimums are maximums for each backend are as follows:

| Backend    | Minimum  | Maximum  |
|------------|----------|----------|
| SQLite 3   | *(does not apply)*  | *(does not apply)* |
| PostgreSQL | 1 second | forever  |
| MySQL      | 1 second | forever  |

Note that in practice neither PostgreSQL nor MySQL will wait indefinitely: they are still subject to PHP's socket timeouts. Consult [PostgreSQL's documentation](https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-CONNECT-CONNECT-TIMEOUT) for details on how the timeout is interpreted by PostgreSQL. 

### dbTimeoutExec

| Type     | Default |
|----------|---------|
| interval | `null`  |

The number of seconds to wait before returning a timeout error when executing a database operation (i.e. computing results). The special value `null` waits the maximum amount of time, while `0.0` waits the minimum amount of time. The minimums are maximums for each backend are as follows:

| Backend    | Minimum       | Maximum  |
|------------|---------------|----------|
| SQLite 3   | *(does not apply)*       | *(does not apply)* |
| PostgreSQL | 1 millisecond | forever  |
| MySQL      | 1 second      | forever  |

With MySQL this timeout only applies to read operations, whereas PostgreSQL will time out write operations as well. 

### dbTimeoutLock

| Type     | Default |
|----------|---------|
| interval | `60.0`  |

The number of seconds to wait before returning a timeout error when acquiring database locks. The special value `null` waits the maximum amount of time, while `0.0` waits the minimum amount of time. The minimums are maximums for each backend are as follows:

| Backend    | Minimum        | Maximum               |
|------------|----------------|-----------------------|
| SQLite 3   | 0 milliseconds | at least 24 days      |
| PostgreSQL | 1 millisecond  | forever               |
| MySQL      | 1 second       | approximately 1 year  |

Note that PostgreSQL counts time spent waiting for locks as part of the above execution timeout. The maximum timeout for SQLite is `PHP_INT_MAX` milliseconds, which on 32-bit systems is just under 25 days, and on 64-bit systems is billions of years.

## Database settings specific to SQLite 3

### dbSQLite3File

| Type           | Default |
|----------------|---------|
| string or null | `null`  |

The full path and file name of SQLite database. The special value `null` evaluates to a file named `"arsse.db"` in the directory where The Arsse is installed.

### dbSQLite3Key

| Type   | Default |
|--------|---------|
| string | `""`    |

The key used to encrypt/decrypt the SQLite database. This is only relevant if using the [SQLite Encryption Extension](https://www.sqlite.org/see/).

## Database settings specific to PostgreSQL

### dbPostgreSQLHost

| Type   | Default |
|--------|---------|
| string | `""`    |

The host name, address, or socket path of the PostgreSQL database server.

Consult [PostgreSQL's documentation](https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-CONNECT-HOST) for more details.

### dbPostgreSQLUser

| Type   | Default   |
|--------|-----------|
| string | `"arsse"` |

The log-in user name for the PostgreSQL database server.

### dbPostgreSQLPass

| Type   | Default |
|--------|---------|
| string | `""`    |

The log-in password for the PostgreSQL database server.

### dbPostgreSQLPort

| Type    | Default |
|---------|---------|
| integer | `5432`  |

The TCP port on which to connect to the PostgreSQL database server, if connecting via TCP.

### dbPostgreSQLDb

| Type   | Default   |
|--------|-----------|
| string | `"arsse"` |

The name of the database used by The Arsse on the PostgreSQL database server.

### dbPostgreSQLSchema

| Type   | Default |
|--------|---------|
| string | `""`    |

The name of the schema used by The Arsse on the PostgreSQL database server. When not set to the empty string, the PostgreSQL search path is modified to consist of the specified schema with a fallback to the public schema.

Consult [PostgreSQL's documentation](https://www.postgresql.org/docs/current/ddl-schemas.html) for more details.

### dbPostgreSQLService

| Type   | Default |
|--------|---------|
| string | `""`    |

A PostgreSQL service file entry to use *instead of* the above configuration; if using a service entry all above PostgreSQL-specific parameters except schema are ignored.

Consult [PostgreSQL's documentation](https://www.postgresql.org/docs/current/libpq-pgservice.html) for more details.

## Database settings specific to MySQL

### dbMySQLHost

| Type   | Default       |
|--------|---------------|
| string | `"localhost"` |

The host name or address of the MySQL database server. The values `"localhost"` and `"."` are special.

Consult [MySQL's documentation](https://dev.mysql.com/doc/refman/8.0/en/connecting.html) for more details.

### dbMySQLUser

| Type   | Default   |
|--------|-----------|
| string | `"arsse"` |

The log-in user name for the MySQL database server.

### dbMySQLPass

| Type   | Default |
|--------|---------|
| string | `""`    |

The log-in password for the MySQL database server.

### dbMySQLPort

| Type    | Default |
|---------|---------|
| integer | `3306`  |

The TCP port on which to connect to the MySQL database server, if connecting via TCP.

### dbMySQLDb

| Type   | Default   |
|--------|-----------|
| string | `"arsse"` |

The name of the database used by The Arsse on the MySQL database server.

### dbMySQLSocket

| Type   | Default |
|--------|---------|
| string | `""`    |

A Unix domain socket or named pipe to use for the MySQL database server when not connecting via TCP.

## User management settings

### userDriver

| Type   | Default      |
|--------|--------------|
| string | `"internal"` |

The user management driver to use. Currently only `"internal"` is available, which stores user IDs and hashed passwords in the configured database.

It is also possible to specify the fully-qualified name of a class which implements the user management driver interface. For example, specifying `"JKingWeb\Arsse\User\Internal\Driver"` would use the internal driver.

### userPreAuth

| Type    | Default |
|---------|---------|
| boolean | `false` |

Whether users are authenticated by the Web server before requests are passed to The Arsse. If set to `true` The Arsse will perform no HTTP-level authentication and assume that the user ID supplied in either the `REMOTE_USER` CGI variable or the `Authorization` HTTP header-field (if `Basic` authentication was used) is authentic.  

For synchronization protocols which implement their own authentication (such as Tiny Tiny RSS), this setting may or may not affect how protocol-level authentication is handled; consult the section on The Arsse's [supported protocols](/en/Supported_Protocols) for more information.

If the user has not previously logged in, an entry is created for them in the database automatically. If the Web server uses `Basic` HTTP authentication and passes along the `Authorization` field, a hash of the user's password will also be stored in The Arsse's database.

### userHTTPAuthRequired

| Type    | Default |
|---------|---------|
| boolean | `false` |

Whether to require successful HTTP authentication before processing API-level authentication, for protocol which implement their own authentication.

### userSessionEnforced

| Type    | Default |
|---------|---------|
| boolean | `true`  |

Whether invalid or expired API session tokens should prevent logging in when HTTP authentication is used, for protocol which implement their own authentication.

### userSessionTimeout

| Type     | Default   |
|----------|-----------|
| interval | `"PT24H"` |

The period of inactivity after which log-in sessions should be considered invalid. Session timeouts should not be made too long, to guard against session hijacking.

### userSessionLifetime

| Type     | Default   |
|----------|-----------|
| interval | `"P7D"`   |

The maximum lifetime of log-in sessions regardless of recent activity. Session lifetimes should not be made too long, to guard against session hijacking.

### userTempPasswordLength

| Type    | Default |
|---------|---------|
| integer | `20`    |

The desired length in characters of randomly-generated user passwords. When [adding users](/en/Command-Line_Functionality), omitting a desired password generates a random one; this setting controls the length of these passwords.

## Newsfeed fetching service settings

### serviceDriver

| Type   | Default        |
|--------|----------------|
| string | `"subprocess"` |

The newsfeed fetching service driver to use. The following values are understood:

- `"serial"`: Fetches newsfeeds and processed them one at a time. This is the slowest method, but is simple and reliable.
- `"subprocess"`: Fetches and processes multiple newsfeeds concurrently by starting a separate process for each newsfeed using PHP's [`popen`](https://php.net/manual/en/function.popen.php) function. This uses more memory and processing power, but takes less total time.

It is also possible to specify the fully-qualified name of a class which implements the service driver interface. For example, specifying `"JKingWeb\Arsse\Service\Serial\Driver"` would use the serial driver.

### serviceFrequency

| Type     | Default  |
|----------|----------|
| interval | `"PT2M"` |

The interval the newsfeed fetching service observes between checks for new articles. Note that requests to foreign servers are not necessarily made at this frequency: each newsfeed is assigned its own time at which to be next retrieved. This setting instead defines the length of time the fetching service will sleep between periods of activity.

Consult "[Newsfeed Refresh Schedule](/en/Installation/Using_the_Daemon#page_Newsfeed_Refresh_Schedule)" for details on how often newsfeeds are fetched.

### serviceQueueWidth

| Type    | Default |
|---------|---------|
| integer | `5`     |

The maximum number of concurrent newsfeed updates to perform, if a concurrent service driver is used. 

### fetchTimeout

| Type     | Default |
|----------|---------|
| interval | `10.0`  |

The maximum number of seconds to wait for data when fetching newsfeeds from foreign servers.

### fetchSizeLimit

| Type    | Default           |
|---------|-------------------|
| integer | `2 * 1024 * 1024` |

The maximum size, in bytes, of data to accept when fetching a newsfeed. Newsfeeds larger than this will be rejected to guard against denial-of-servioce attacks.

The default value is equal to two megabytes.

### fetchEnableScraping

| Type    | Default |
|---------|---------|
| boolean | `true`  |


CHANGELOG
Version 0.8.1 (2019-??-??)
==========================

Bug fixes:
- Don't crash updating feeds cached solely via ETag

Version 0.8.0 (2019-07-26)
==========================

New features:
- Support for the Fever protocol (see README.md for details)
- Command line functionality for clearing a password, disabling the account
- Command line options for dealing with Fever passwords
- Command line functionality for importing and exporting OPML
- Command line documentation of all commands and options

Bug fixes:
- Treat command line option -h the same as --help
- Sort Tiny Tiny RSS special feeds according to special ordering
- Invalidate sessions when passwords are changed
- Correct example systemd unit to start after PostgreSQL and MySQL

Changes:
- Perform regular database maintenance to improve long-term performance

Version 0.7.1 (2019-03-25)
==========================

Bug fixes:
- Correctly initialize new on-disk SQLite databases
- Retry queries on schema changes with PDO SQLite
- Correctly read author name from database in Tiny Tiny RSS
- Update internal version number to correct version

Changes:
- Improve performance of lesser-used database queries

Version 0.7.0 (2019-03-02)
==========================

New features:
- Support for basic freeform searching in Tiny Tiny RSS
- Console command to refresh all stale feeds once then exit

Bug fixes:
- Ensure updating does not fail with newsfeeds larger than 250 entries

Version 0.6.1 (2019-01-23)
==========================

Bug Fixes:
- Unify SQL timeout settings
- Correctly escape shell command in subprocess service driver
- Correctly allow null time intervals in configuration when appropriate

Changes:
- Change PicoFeed dependency to maintained version (Thanks, Aaron Parecki!)
- Remove non-functional cURL service driver

Version 0.6.0 (2019-01-21)
==========================

New features:
- Support for PostgreSQL databases
- Support for MySQL databases
- Validation of configuration parameters

Bug fixes:
- Use a general-purpose Unicode collation with SQLite databases
- Use the correct SQLite schema change procedure for 3.25 and later

Changes:
- Improve performance of common database queries by 80-90%
- Make configuration defaults consistent with their defined types

Version 0.5.1 (2018-11-10)
==========================

Bug fixes:
- Correctly initialize PDO database driver

Version 0.5.0 (2018-11-07)
==========================

New features:
- Command line functionality for listing, removing, and changing the password
  of users as well as testing authentication

Bug fixes:
- Print command-line error messages more sensibly
- Allow exporting default configuration to standard output
- Fail correctly on authentication failure
- Prefer JSON data over GET parameters in NextCloud News

Changes:
- Simplify user management backend to minimize opportunity for bugs
- Document previously unknown NextCloud News behaviour

Version 0.4.0 (2018-10-26)
==========================

New features:
- Support for HTTP authentication in Tiny Tiny RSS (see README.md for details)
- New userHTTPAuthRequired and userSessionEnforced settings

Version 0.3.1 (2018-07-22)
==========================

Bug fixes:
- Correctly read the configuration file on start-up
- Minor fixes to code and documentation

Changes:
- Disable memory and time limits to avoid deadlocks with NextCloud News

Version 0.3.0 (2018-01-12)
==========================

New features:
- Support for SQLite3 via PDO
- Support for cross-origin resource sharing in all protocols

Bug fixes:
- Correctly handle %-encoded request URLs
- Overhaul protocol detection to fix various subtle bugs
- Overhaul HTTP response handling for more consistent results

Changes:
- Make date strings in TTRSS explicitly UTC

Version 0.2.1 (2017-12-07)
==========================

Bug fixes:
- Make SQLite timeouts much less likely
- Accept base64-encoded passwords from TTRSS clients
- Rename feeds correctly via TTRSS protocol
- Toggle marks correctly via TTRSS protocol
- Sort everything case-insensitively
- Be even stricter about output data types in NextCloud News

Changes:
- Do not omit read feeds from TTRSS' getCounters, to fix some clients

Version 0.2.0 (2017-11-30)
==========================

New features:
- Support for the Tiny Tiny RSS protocol (see README.md for details)
- Support for HTTP OPTIONS requests in all protocols

Bug fixes:
- Perform feed discovery *correctly*
- Expose the incorrectDbCharset boolean in the NextCloud News server status
- Give NextCloud News articles' guidHash attribute the correct type (string)

Changes:
- Overhaul input type normalization to minimize bug opportunities

Version 0.1.1 (2017-09-30)
==========================

Bug fixes:
- Perform feed discovery like NextCloud News does
- Respond correctly to HEAD requests
- Various minor fixes

Version 0.1.0 (2017-08-29)
==========================

Initial release


lib/Conf.php
<?php
/** @license MIT
 * Copyright 2017 J. King, Dustin Wilson et al.
 * See LICENSE and AUTHORS files for details */

/** Conf class */
declare(strict_types=1);
namespace JKingWeb\Arsse;

use JKingWeb\Arsse\Misc\ValueInfo as Value;

/** Class for loading, saving, and querying configuration
 *
 * The Conf class serves both as a means of importing and querying configuration information, as well as a source for default parameters when a configuration file does not specify a value.
 * All public properties are configuration parameters that may be set by the server administrator. */
class Conf {
    /** @var string Default language to use for logging and errors */
    public $lang                    = "en";

    /** @var string The database driver to use, one of "sqlite3", "postgresql", or "mysql". A fully-qualified class name may also be used for custom drivers */
    public $dbDriver                = "sqlite3";
    /** @var boolean Whether to attempt to automatically update the database when upgrading to a new version with schema changes */
    public $dbAutoUpdate            = true;
    /** @var \DateInterval|null Number of seconds to wait before returning a timeout error when connecting to a database (null waits forever; not applicable to SQLite) */
    public $dbTimeoutConnect        = 5.0;
    /** @var \DateInterval|null Number of seconds to wait before returning a timeout error when executing a database operation (null waits forever; not applicable to SQLite) */
    public $dbTimeoutExec           = null;
    /** @var \DateInterval|null Number of seconds to wait before returning a timeout error when acquiring a database lock (null waits forever) */
    public $dbTimeoutLock           = 60.0;
    /** @var string|null Full path and file name of SQLite database (if using SQLite) */
    public $dbSQLite3File           = null;
    /** @var string Encryption key to use for SQLite database (if using a version of SQLite with SEE) */
    public $dbSQLite3Key            = "";
    /** @var string Host name, address, or socket path of PostgreSQL database server (if using PostgreSQL) */
    public $dbPostgreSQLHost        = "";
    /** @var string Log-in user name for PostgreSQL database server (if using PostgreSQL) */
    public $dbPostgreSQLUser        = "arsse";
    /** @var string Log-in password for PostgreSQL database server (if using PostgreSQL) */
    public $dbPostgreSQLPass        = "";
    /** @var integer Listening port for PostgreSQL database server (if using PostgreSQL over TCP) */
    public $dbPostgreSQLPort        = 5432;
    /** @var string Database name on PostgreSQL database server (if using PostgreSQL) */
    public $dbPostgreSQLDb          = "arsse";
    /** @var string Schema name in PostgreSQL database (if using PostgreSQL) */
    public $dbPostgreSQLSchema      = "";
    /** @var string Service file entry to use (if using PostgreSQL); if using a service entry all above parameters except schema are ignored */
    public $dbPostgreSQLService     = "";
    /** @var string Host name or address of MySQL database server (if using MySQL) */
    public $dbMySQLHost             = "localhost";
    /** @var string Log-in user name for MySQL database server (if using MySQL) */
    public $dbMySQLUser             = "arsse";
    /** @var string Log-in password for MySQL database server (if using MySQL) */
    public $dbMySQLPass             = "";
    /** @var integer Listening port for MySQL database server (if using MySQL over TCP) */
    public $dbMySQLPort             = 3306;
    /** @var string Database name on MySQL database server (if using MySQL) */
    public $dbMySQLDb               = "arsse";
    /** @var string Unix domain socket or named pipe to use for MySQL when not connecting over TCP */
    public $dbMySQLSocket           = "";

    /** @var string The user management driver to use, currently only "internal". A fully-qualified class name may also be used for custom drivers */
    public $userDriver              = "internal";
    /** @var boolean Whether users are already authenticated by the Web server before the application is executed */
    public $userPreAuth             = false;
    /** @var boolean Whether to require successful HTTP authentication before processing API-level authentication for protocols which have any. Normally the Tiny Tiny RSS relies on its own session-token authentication scheme, for example */
    public $userHTTPAuthRequired    = false;
    /** @var integer Desired length of temporary user passwords */
    public $userTempPasswordLength  = 20;
    /** @var boolean Whether invalid or expired API session tokens should prevent logging in when HTTP authentication is used, for protocol which implement their own authentication */
    public $userSessionEnforced     = true;
    /** @var \DateInterval Period of inactivity after which log-in sessions should be considered invalid, as an ISO 8601 duration (default: 24 hours)
     * @see https://en.wikipedia.org/wiki/ISO_8601#Durations */
    public $userSessionTimeout      = "PT24H";
    /** @var \DateInterval Maximum lifetime of log-in sessions regardless of activity, as an ISO 8601 duration (default: 7 days);
     * @see https://en.wikipedia.org/wiki/ISO_8601#Durations */
    public $userSessionLifetime     = "P7D";

    /** @var string Feed update service driver to use, one of "serial" or "subprocess". A fully-qualified class name may also be used for custom drivers */
    public $serviceDriver           = "subprocess";
    /** @var \DateInterval The interval between checks for new articles, as an ISO 8601 duration
     * @see https://en.wikipedia.org/wiki/ISO_8601#Durations */
    public $serviceFrequency        = "PT2M";
    /** @var integer Number of concurrent feed updates to perform */
    public $serviceQueueWidth       = 5;

    /** @var \DateInterval Number of seconds to wait for data when fetching feeds from foreign servers */
    public $fetchTimeout            = 10.0;
    /** @var integer Maximum size, in bytes, of data when fetching feeds from foreign servers */
    public $fetchSizeLimit          = 2 * 1024 * 1024;
    /** @var boolean Whether to allow the possibility of fetching full article contents using an item's URL. Whether fetching will actually happen is also governed by a per-feed setting */
    public $fetchEnableScraping     = true;
    /** @var string|null User-Agent string to use when fetching feeds from foreign servers */
    public $fetchUserAgentString    = null;

    /** @var \DateInterval|null When to delete a feed from the database after all its subscriptions have been deleted, as an ISO 8601 duration (default: 24 hours; null for never)
     * @see https://en.wikipedia.org/wiki/ISO_8601#Durations */
    public $purgeFeeds              = "PT24H";
    /** @var \DateInterval|null When to delete an unstarred article in the database after it has been marked read by all users, as an ISO 8601 duration (default: 7 days; null for never)
     * @see https://en.wikipedia.org/wiki/ISO_8601#Durations */
    public $purgeArticlesRead       = "P7D";
    /** @var \DateInterval|null When to delete an unstarred article in the database regardless of its read state, as an ISO 8601 duration (default: 21 days; null for never)
     * @see https://en.wikipedia.org/wiki/ISO_8601#Durations */
    public $purgeArticlesUnread     = "P21D";

    /** @var string Application name to present to clients during authentication */
    public $httpRealm               = "The Advanced RSS Environment";
    /** @var string Space-separated list of origins from which to allow cross-origin resource sharing */
    public $httpOriginsAllowed      = "*";
    /** @var string Space-separated list of origins from which to deny cross-origin resource sharing */
    public $httpOriginsDenied       = "";

    ### OBSOLETE SETTINGS

    /** @var \DateInterval|null (OBSOLETE) Number of seconds for SQLite to wait before returning a timeout error when trying to acquire a write lock on the database (zero does not wait) */
    public $dbSQLite3Timeout        = null; // previously 60.0

    const TYPE_NAMES = [
        Value::T_BOOL     => "boolean",
        Value::T_STRING   => "string",
        Value::T_FLOAT    => "float",
        VALUE::T_INT      => "integer",
        Value::T_INTERVAL => "interval",
    ];
    const EXPECTED_TYPES = [
        'dbTimeoutExec'    => "double",
        'dbTimeoutLock'    => "double",
        'dbTimeoutConnect' => "double",
        'dbSQLite3Timeout' => "double",
    ];

    protected static $types = [];

    /** Creates a new configuration object
     * @param string $import_file Optional file to read configuration data from
     * @see self::importFile() */
    public function __construct(string $import_file = "") {
        if (!static::$types) {
            static::$types = $this->propertyDiscover();
        }
        foreach (array_keys(static::$types) as $prop) {
            $this->$prop = $this->propertyImport($prop, $this->$prop);
        }
        if ($import_file !== "") {
            $this->importFile($import_file);
        }
    }

    /** Layers configuration data from a file into an existing object
     *
     * The file must be a PHP script which returns an array with keys that match the properties of the Conf class. Malformed files will throw an exception; unknown keys are silently accepted. Files may be imported in succession, though this is not currently used.
     * @param string $file Full path and file name for the file to import */
    public function importFile(string $file): self {
        if (!file_exists($file)) {
            throw new Conf\Exception("fileMissing", $file);
        } elseif (!is_readable($file)) {
            throw new Conf\Exception("fileUnreadable", $file);
        }
        try {
            ob_start();
            $arr = (@include $file);
        } catch (\Throwable $e) {
            $arr = null;
        } finally {
            ob_end_clean();
        }
        if (!is_array($arr)) {
            throw new Conf\Exception("fileCorrupt", $file);
        }
        return $this->importData($arr, $file);
    }

    /** Layers configuration data from an associative array into an existing object
     *
     * The input array must have keys that match the properties of the Conf class; unknown keys are silently accepted. Arrays may be imported in succession, though this is not currently used.
     * @param mixed[] $arr Array of configuration parameters to export */
    public function import(array $arr): self {
        $file = debug_backtrace(\DEBUG_BACKTRACE_IGNORE_ARGS, 1)[0]['file'] ?? "";
        return $this->importData($arr, $file);
    }

    /** Layers configuration data from an associative array into an existing object */
    protected function importData(array $arr, string $file): self {
        foreach ($arr as $key => $value) {
            $this->$key = $this->propertyImport($key, $value, $file);
        }
        return $this;
    }

    /** Outputs configuration settings, either non-default ones or all, as an associative array
     * @param bool $full Whether to output all configuration options rather than only changed ones */
    public function export(bool $full = false): array {
        $conf = new \ReflectionObject($this);
        $ref = (new \ReflectionClass($this))->getDefaultProperties();
        $out = [];
        foreach ($conf->getProperties(\ReflectionProperty::IS_PUBLIC) as $prop) {
            $name = $prop->name;
            $value = $prop->getValue($this);
            if ($prop->isDefault()) {
                $default = $ref[$name];
                // if the property is a known property (rather than one added by a hypothetical plug-in)
                // we convert intervals to strings and then export anything which doesn't match the default value
                $value = $this->propertyExport($name, $value);
                if ((is_scalar($value) || is_null($value)) && ($full || $value !== $ref[$name])) {
                    $out[$name] = $value;
                }
            } elseif (is_scalar($value) || is_null($value)) {
                // otherwise export the property only if it is scalar
                $out[$name] = $value;
            }
        }
        return $out;
    }

    /** Outputs configuration settings, either non-default ones or all, to a file in a format suitable for later import
     * @param string $file Full path and file name for the file to import to; the containing directory must already exist
     * @param bool $full Whether to output all configuration options rather than only changed ones */
    public function exportFile(string $file, bool $full = false): bool {
        $arr = $this->export($full);
        $conf = new \ReflectionObject($this);
        $out = "<?php return [".PHP_EOL;
        foreach ($arr as $prop => $value) {
            $match = null;
            $doc = $comment = "";
            // retrieve the property's docblock, if it exists
            try {
                $doc = (new \ReflectionProperty(self::class, $prop))->getDocComment();
                // parse the docblock to extract the property description
                if (preg_match("<@var\s+\S+\s+(.+?)(?:\s*\*/)?\s*$>m", $doc, $match)) {
                    $comment = $match[1];
                }
            } catch (\ReflectionException $e) {
            }
            // append the docblock description if there is one, or an empty comment otherwise
            $out .= " // ".$comment.PHP_EOL;
            // append the property and an export of its value to the output
            $out .= " ".var_export($prop, true)." => ".var_export($value, true).",".PHP_EOL;
        }
        $out .= "];".PHP_EOL;
        // write the configuration representation to the requested file
        if (!@file_put_contents($file, $out)) {
            // if it fails throw an exception
            $err = file_exists($file) ? "fileUnwritable" : "fileUncreatable";
            throw new Conf\Exception($err, $file);
        }
        return true;
    }

    /** Caches information about configuration properties for later access */
    protected function propertyDiscover(): array {
        $out = [];
        $rc = new \ReflectionClass($this);
        foreach ($rc->getProperties(\ReflectionProperty::IS_PUBLIC) as $p) {
            if (preg_match("/@var\s+((?:int(eger)?|float|bool(ean)?|string|\\\\DateInterval)(?:\|null)?)[^\[]/", $p->getDocComment(), $match)) {
                $match = explode("|", $match[1]);
                $nullable = (sizeof($match) > 1);
                $type = [
                    'string'         => Value::T_STRING   | Value::M_STRICT,
                    'integer'        => Value::T_INT      | Value::M_STRICT,
                    'boolean'        => Value::T_BOOL     | Value::M_STRICT,
                    'float'          => Value::T_FLOAT    | Value::M_STRICT,
                    '\\DateInterval' => Value::T_INTERVAL | Value::M_LOOSE,
                ][$match[0]];
                if ($nullable) {
                    $type |= Value::M_NULL;
                }
            } else {
                $type = Value::T_MIXED; // @codeCoverageIgnore
            }
            $out[$p->name] = ['name' => $match[0], 'const' => $type];
        }
        return $out;
    }

    protected function propertyImport(string $key, $value, string $file = "") {
        $typeName = static::$types[$key]['name'] ?? "mixed";
        $typeConst = static::$types[$key]['const'] ?? Value::T_MIXED;
        $nullable = (int) (bool) (static::$types[$key]['const'] & Value::M_NULL);
        try {
            if ($typeName === "\\DateInterval") {
                // date intervals have special handling: if the existing value (ultimately, the default value)
                // is an integer or float, the new value should be imported as numeric. If the new value is a string
                // it is first converted to an interval and then converted to the numeric type if necessary
                $mode = $nullable ? Value::M_STRICT | Value::M_NULL : Value::M_STRICT;
                if (is_string($value)) {
                    $value =  Value::normalize($value, Value::T_INTERVAL | $mode);
                }
                switch (self::EXPECTED_TYPES[$key] ?? gettype($this->$key)) {
                    case "integer":
                        return Value::normalize($value, Value::T_INT | $mode); // @codeCoverageIgnore
                    case "double":
                        return Value::normalize($value, Value::T_FLOAT | $mode);
                    case "string":
                    case "object":
                        return $value;
                    default:
                        throw new Conf\Exception("ambiguousDefault", ['param' => $key]); // @codeCoverageIgnore
                }
            }
            $value =  Value::normalize($value, $typeConst);
            switch ($key) {
                case "dbDriver":
                    $driver = $driver ?? Database::DRIVER_NAMES[strtolower($value)] ?? $value;
                    $interface = $interface ?? Db\Driver::class;
                    // no break
                case "userDriver":
                    $driver = $driver ?? User::DRIVER_NAMES[strtolower($value)] ?? $value;
                    $interface = $interface ?? User\Driver::class;
                    // no break
                case "serviceDriver":
                    $driver = $driver ?? Service::DRIVER_NAMES[strtolower($value)] ?? $value;
                    $interface = $interface ?? Service\Driver::class;
                    if (!is_subclass_of($driver, $interface)) {
                        throw new Conf\Exception("semanticMismatch", ['param' => $key, 'file' => $file]);
                    }
                    return $driver;
            }
            return $value;
        } catch (ExceptionType $e) {
            $type =  static::$types[$key]['const'] & ~(Value::M_STRICT | Value::M_DROP | Value::M_NULL | Value::M_ARRAY);
            throw new Conf\Exception("typeMismatch", ['param' => $key, 'type' => self::TYPE_NAMES[$type], 'file' => $file, 'nullable' => $nullable]);
        }
    }

    protected function propertyExport(string $key, $value) {
        $value = ($value instanceof \DateInterval) ? Value::normalize($value, Value::T_STRING) : $value;
        switch ($key) {
            case "dbDriver":
                return array_flip(Database::DRIVER_NAMES)[$value] ?? $value;
            case "userDriver":
                return array_flip(User::DRIVER_NAMES)[$value] ?? $value;
            case "serviceDriver":
                return array_flip(Service::DRIVER_NAMES)[$value] ?? $value;
            default:
                return $value;
        }
    }
}
